sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/reflect/DoFnSignature.java

```diff
@@ -114,11 +114,116 @@ static Builder builder() {
   }
 
   /** A type of optional parameter of the {@link DoFn.ProcessElement} method. */
-  public enum Parameter {
-    BOUNDED_WINDOW,
-    INPUT_PROVIDER,
-    OUTPUT_RECEIVER,
-    RESTRICTION_TRACKER
+  public static abstract class Parameter {
+
+    public interface Visitor<ResultT> {
+      ResultT visit(BoundedWindowParameter p);
+      ResultT visit(InputProviderParameter p);
+      ResultT visit(OutputReceiverParameter p);
+      ResultT visit(RestrictionTrackerParameter p);
+      ResultT visit(StateParameter p);
+
+      /**
+       * A base class for a visitor with a default method for cases it is not interested in.
+       */
+      public static abstract class Defaults<ResultT> implements Visitor<ResultT> {
+
+        protected abstract ResultT visitDefault(Parameter p);
+
+        @Override
+        public ResultT visit(BoundedWindowParameter p) {
+          return visitDefault(p);
+        }
+
+        @Override
+        public ResultT visit(InputProviderParameter p) {
+          return visitDefault(p);
+        }
+
+        @Override
+        public ResultT visit(OutputReceiverParameter p) {
+          return visitDefault(p);
+        }
+
+        @Override
+        public ResultT visit(RestrictionTrackerParameter p) {
+          return visitDefault(p);
+        }
+
+        @Override
+        public ResultT visit(StateParameter p) {
+          return visitDefault(p);
+        }
+      }
+    }
+
+    public abstract <ResultT> ResultT accept(Visitor<ResultT> visitor);
+
+    public static BoundedWindowParameter boundedWindow() {
```

Comment:
These could have stayed constants: BOUNDED_WINDOW = new ..., which would make it less awkward to read "contains(boundedWindow())" ("are you sure it implements equality?")